### DISCUSSION TITLE
`cy.visit("/{PATH}")` returns 404 while the site is completely reachable and I don't see a 404.

### DISCUSSION THREAD

**Discussion by hamidmayeli** [cypress-io/cypress#30729] [Questions and Help]
I am testing a ViteJS + React. The test works fine. But I like to upgrade to react 19 and I am facing with 404 when visiting `cy.visit("/login")` or other pages. The repo is based on this [template](https://github.com/hamidmayeli/quick-spa).

Here is the config
```ts
import { defineConfig } from "cypress";
import registerCodeCoverageTasks from "@cypress/code-coverage/task";

export default defineConfig({
  e2e: {
    viewportWidth: 960,
    screenshotOnRunFailure: true,
    video: true,
    retries: {
      runMode: 2,
    },
    baseUrl: "http://localhost:5173",
    setupNodeEvents(on, config) {
      registerCodeCoverageTasks(on, config);
      // implement node event listeners here

      return config;
    },
  },

  component: {
    viewportWidth: 960,
    screenshotOnRunFailure: false,
    retries: {
      runMode: 2,
    },
    video: false,
    devServer: {
      framework: "react",
      bundler: "vite",
    },
    setupNodeEvents(on, config) {
      registerCodeCoverageTasks(on, config);
      // implement node event listeners here

      return config;
    },
  },
});
```

and here is a sample test
```ts
describe("App", () => {
  it("can log in and out", () => {
    cy.visit("/");

    cy.visit("/login");
....
```

I also see a few error logs like `queue.insert must be called with a valid index - the index (9) is out of bounds` or `(uncaught exception)TypeError: Cannot read properties of undefined (reading '_fired')`

**Comment by portyu9** [comment]
Hi @hamidmayeli 

A React 19 upgrade by itself should **not make `cy.visit('/login')` return HTTP 404**.

For E2E testing, this:

```js
cy.visit('/login')
```

makes Cypress request:

```text
http://localhost:5173/login
```

because that is your configured `baseUrl`.

Cypress expects the navigation response to be successful HTML. So the first thing I would check is the actual server response independently of React Router:

```bash
curl -i http://localhost:5173/login
```

Also try opening that exact URL in a **fresh/private browser profile**, rather than navigating to `/login` from `/` using the application's client-side router.

If that request really returns `404`, then the server needs to be configured to serve your SPA's `index.html` for application routes such as `/login`. Setting:

```js
failOnStatusCode: false
```

would only hide the underlying server-routing problem and would not be the correct fix.

There is also an important version change since this was posted: **Cypress 14 added official React 19 Component Testing support**. Current Cypress supports React 18 and 19.

So if the React 19 upgrade was also being used with Component Testing on Cypress 13 or earlier, upgrade Cypress to a version with React 19 support.

The internal errors:

```text
queue.insert must be called with a valid index
```

and:

```text
Cannot read properties of undefined (reading '_fired')
```

should also be retested after upgrading Cypress, because those are separate from an HTTP 404.

I would therefore debug these as two issues:

1. **`cy.visit('/login')` 404** → verify the direct Vite/server response and SPA fallback
2. **internal Cypress errors after React 19** → upgrade to a React-19-supported Cypress version and reproduce

Do not treat the 404 as a React Router assertion failure—`cy.visit()` is failing earlier at the HTTP navigation level.

Please mark this answer as accepted if it helped, thank you.
